builds.go

```diff
@@ -0,0 +1,293 @@
+		return nil, fmt.Errorf("%s missing %s arch", img, arch)
+	}
+	// TODO warn/error on multiple entries for arch? (would mean something like index/manifest list with multiple os.version values for Windows - we avoid this in DOI today, but we don't have any automated *checks* for it, so the current state is a little precarious)
+
+	ref := func(obj *registry.ResolvedObject) string {
+		base, _, _ := strings.Cut(obj.ImageRef, "@")
+		base = strings.TrimPrefix(base, "docker.io/")
+		base = strings.TrimPrefix(base, "library/")
+		return base + "@" + string(obj.Desc.Digest)
+	}
+	resolved := &RemoteResolvedFull{
+		Manifest: RemoteResolved{
+			Ref:  ref(&rArches[arch][0]),
+			Desc: rArches[arch][0].Desc,
+		},
+	}
+	if r.IsImageIndex() {
+		resolved.Index = &RemoteResolved{
+			Ref:  ref(r),
+			Desc: r.Desc,
+		}
+	}
+	return resolved, nil
+}
+
+func main() {
+	sourcesJsonFile := os.Args[1] // "sources.json"
+
+	stagingTemplate := os.Getenv("BASHBREW_STAGING_TEMPLATE") // "oisupport/staging-ARCH:BUILD"
+	if !strings.Contains(stagingTemplate, "BUILD") {
+		panic("invalid BASHBREW_STAGING_TEMPLATE (missing BUILD)")
+	}
+
+	// containerd uses logrus, but it defaults to "info" (which is a bit leaky where we use containerd)
+	logrus.SetLevel(logrus.WarnLevel)
+
+	type out struct {
+		buildId string
+		json    []byte
+	}
+	outs := make(chan chan out, concurrency) // we want the end result to be "in order", so we have a channel of channels of outputs so each output can be generated async (and write to the "inner" channel) and the outer channel stays in the input order
+
+	go func() {
+		// Go does not have ordered maps *and* is complicated to read an object, make a tiny modification, write it back out (without modelling the entire schema), so we'll let a single invocation of jq solve both problems (munging the documents in the way we expect *and* giving us an in-order stream)
+		jq := exec.Command("jq", "-c", ".[] | (.arches | to_entries[]) as $arch | .arches = { ($arch.key): $arch.value }", sourcesJsonFile)
+		jq.Stderr = os.Stderr
+
+		stdout, err := jq.StdoutPipe()
+		if err != nil {
+			panic(err)
+		}
+		if err := jq.Start(); err != nil {
+			panic(err)
+		}
+
+		sourceArchResolved := map[string](func() *RemoteResolvedFull){}
+		sourceArchResolvedMutex := sync.RWMutex{}
+
+		decoder := json.NewDecoder(stdout)
+		for decoder.More() {
+			var build MetaBuild
+
+			if err := decoder.Decode(&build.Source); err == io.EOF {
+				break
+			} else if err != nil {
+				panic(err)
+			}
+
+			var source MetaSource
+			if err := json.Unmarshal(build.Source, &source); err != nil {
+				panic(err)
+			}
+
+			build.Build.SourceID = source.SourceID
+
+			if len(source.Arches) != 1 {
+				panic("unexpected arches length: " + string(build.Source))
+			}
+			for build.Build.Arch = range source.Arches {
+				// I really hate Go.
+				// (just doing a lookup of the only key in my map into a variable)
```

Comment:
Perhaps soon, https://pkg.go.dev/golang.org/x/exp/maps#Keys

Reply:
Indeed!  I was angry that https://pkg.go.dev/maps didn't include that (yet), but perhaps there is still hope.

Reply:
Ahh, those functions being included in stdlib is blocked on the "iterators" proposals (which are blocked on "`for range` over a function" that was accepted but still experimental and requires `GODEBUG` incantations to use), but an iterator would honestly be even better for this use case than a slice of keys so I guess I'll sit down and try to be patient. :joy: